src/lib/optimizer/strategy/join_to_predicate_rewrite_rule.cpp

```diff
@@ -0,0 +1,187 @@
+#include "join_to_predicate_rewrite_rule.hpp"
+
+#include <magic_enum.hpp>
+
+#include "expression/abstract_expression.hpp"
+#include "expression/binary_predicate_expression.hpp"
+#include "expression/expression_functional.hpp"
+#include "expression/expression_utils.hpp"
+#include "expression/lqp_column_expression.hpp"
+#include "logical_query_plan/lqp_utils.hpp"
+#include "logical_query_plan/predicate_node.hpp"
+#include "logical_query_plan/projection_node.hpp"
+
+using namespace hyrise::expression_functional;  // NOLINT
+
+namespace hyrise {
+
+std::string JoinToPredicateRewriteRule::name() const {
+  static const auto name = std::string{"JoinToPredicateRewriteRule"};
+  return name;
+}
+
+void JoinToPredicateRewriteRule::_apply_to_plan_without_subqueries(
+    const std::shared_ptr<AbstractLQPNode>& lqp_root) const {
+  // rewritables finally contains all rewritable join nodes, there unused input side and the predicate to be used for
+  // the rewrite.
+  auto rewritables = std::vector<std::tuple<std::shared_ptr<JoinNode>, LQPInputSide, std::shared_ptr<PredicateNode>>>{};
+
+  visit_lqp(lqp_root, [&](const auto& node) {
+    if (node->type == LQPNodeType::Join) {
+      const auto join_node = std::static_pointer_cast<JoinNode>(node);
+      _gather_rewrite_info(join_node, rewritables);
+    }
+    return LQPVisitation::VisitInputs;
+  });
+
+  const auto rewritables_size = rewritables.size();
+  for (auto index = size_t{0}; index < rewritables_size; ++index) {
+    _perform_rewrite(std::get<0>(rewritables[index]), std::get<1>(rewritables[index]), std::get<2>(rewritables[index]));
+  }
+}
+
+void JoinToPredicateRewriteRule::_gather_rewrite_info(
+    const std::shared_ptr<JoinNode>& join_node,
+    std::vector<std::tuple<std::shared_ptr<JoinNode>, LQPInputSide, std::shared_ptr<PredicateNode>>>& rewritables) {
+  const auto removable_side = join_node->get_unused_input();
+  if (!removable_side.has_value()) {
+    return;
+  }
+
+  std::shared_ptr<AbstractLQPNode> removable_subtree = nullptr;
+  std::shared_ptr<PredicateNode> valid_predicate = nullptr;
+
+  removable_subtree = join_node->input(*removable_side);
+
+  // We don't handle the more complicated case of multiple join predicates. Rewriting in these cases would require
+  // finding and combining multiple suitable predicates. We also don't rewrite anti-joins.
+  if (join_node->join_predicates().size() != 1 || join_node->join_mode == JoinMode::AntiNullAsTrue ||
```

Comment:
Why not `predicate_count > 1`?

Reply:
We cannot rewrite joins with no predicate, so the check should be fine here.

Reply:
Ahhh, it's an early out. Misread that. Got it!